src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs

```diff
@@ -453,6 +453,28 @@ private void OnExecutionMessageReceived(object sender, MessageReceivedEventArgs
 
                         this.channel.Send(data);
                         break;
+
+                    case MessageType.AttachDebuggerToProcess:
+#pragma warning disable SA1119 // Statement must not use unnecessary parenthesis
+                        if (!(testRunEventsHandler is ITestRunEventsHandler2 handler))
+#pragma warning restore SA1119 // Statement must not use unnecessary parenthesis
+                        {
+                            throw new NotSupportedException(string.Format(
+                                CultureInfo.CurrentUICulture,
+                                "AttachDebuggerToProcess is only supported in ITestRunEventsHandler2, but it was called with {0}.",
```

Comment:
ITestRunEventsHandler is implemented in a number of places including in VS Test Window (See TestPlatformProvider.cs). Its a bit strange to see an expectation that only the newer interface should be passed here. Are we certain that the old interface will never be passed to this function for any scenario?

Reply:
The expectation of having only an `ITestRunEventsHandler2` as opposed to having `ITestRunEventsHandler` instead comes with the message type.

We shall receive a `AttachDebuggerToProcess` message from the `testhost` either if the adapter calls its corresponding `AttachDebuggerToProcess` method via the `IFrameworkHandle2`, or if we deemed it necessary to attach to the default `testhost`. Either way, we should be getting an `ITestRunEventsHandler2` instance. If we don't, it's an application error that we should be taking care of that.

Took care of the exception/return false inconsistency. Removed the exception throwing altogether and instead decided to rely on an unconditional cast. If that cast fails and throws an invalid cast exception we should investigate that further as a logic error (we shouldn't be getting in that state in the first place).